OrbitCore/LinuxPerfUtils.cpp

```diff
@@ -27,138 +28,111 @@ perf_event_attr LinuxPerfUtils::generic_perf_event_attr() {
-
+int32_t generic_event_open(perf_event_attr* attr, pid_t pid, int32_t cpu) {
+  int32_t fd = perf_event_open(attr, pid, cpu, -1, 0);
   if (fd == -1) {
     PRINT("perf_event_open error: %s\n", strerror(errno));
   }
-
   return fd;
 }
+}  // namespace
+}  // namespace LinuxPerfUtils
 
-//-----------------------------------------------------------------------------
-int32_t LinuxPerfUtils::context_switch_open(pid_t a_PID, int32_t a_CPU) {
-  perf_event_attr pe = generic_perf_event_attr();
+int32_t LinuxPerfUtils::task_event_open(int32_t cpu) {
+  perf_event_attr pe = generic_event_attr();
   pe.type = PERF_TYPE_SOFTWARE;
   pe.config = PERF_COUNT_SW_DUMMY;
-  pe.context_switch = 1;
+  pe.task = 1;
 
-  int32_t fd = perf_event_open(&pe, a_PID, a_CPU, -1 /*group_fd*/, 0 /*flags*/);
+  return generic_event_open(&pe, -1, cpu);
+}
 
-  if (fd == -1) {
-    PRINT("perf_event_open error: %s\n", strerror(errno));
-  }
+int32_t LinuxPerfUtils::context_switch_open(pid_t pid, int32_t cpu) {
+  perf_event_attr pe = generic_event_attr();
+  pe.type = PERF_TYPE_SOFTWARE;
+  pe.config = PERF_COUNT_SW_DUMMY;
+  pe.context_switch = 1;
 
-  return fd;
+  return generic_event_open(&pe, pid, cpu);
 }
 
-//-----------------------------------------------------------------------------
-int32_t LinuxPerfUtils::stack_sample_event_open(pid_t a_PID,
-                                                uint64_t a_Frequency) {
-  perf_event_attr pe = generic_perf_event_attr();
+int32_t LinuxPerfUtils::stack_sample_event_open(pid_t pid, uint64_t period_ns) {
+  perf_event_attr pe = generic_event_attr();
   pe.type = PERF_TYPE_SOFTWARE;
   pe.config = PERF_COUNT_SW_CPU_CLOCK;
-  pe.sample_freq = a_Frequency;
-  pe.freq = 1;
+  pe.sample_period = period_ns;
   pe.sample_type |= PERF_SAMPLE_STACK_USER | PERF_SAMPLE_REGS_USER;
   pe.task = 1;
   pe.mmap = 1;
 
-  int32_t fd = perf_event_open(&pe, a_PID, -1, -1, 0);
-
-  if (fd == -1) {
-    PRINT("perf_event_open error: %s\n", strerror(errno));
-  }
-
-  return fd;
+  return generic_event_open(&pe, pid, -1);
 }
 
-//-----------------------------------------------------------------------------
 bool LinuxPerfUtils::supports_perf_event_uprobes() {
   return LinuxUtils::GetKernelVersion() >= KERNEL_VERSION(4, 17, 0);
 }
 
-//-----------------------------------------------------------------------------
-int32_t LinuxPerfUtils::uprobe_event_open(const char* a_Module,
-                                          uint64_t a_FunctionOffset,
-                                          pid_t a_PID, int32_t a_CPU,
-                                          int32_t a_GroupFd,
-                                          uint64_t additonal_sample_type) {
-  SCOPE_TIMER_FUNC;
-  PRINT_VAR(a_Module);
-  PRINT_VAR(a_FunctionOffset);
-  PRINT_VAR(a_PID);
-  PRINT_VAR(a_CPU);
-  PRINT_VAR(a_GroupFd);
-  PRINT_VAR(additonal_sample_type);
-
-  perf_event_attr pe = generic_perf_event_attr();
+namespace LinuxPerfUtils {
```

Comment:
Is the anonymous namespace inside the LinuxPerfUtils namespace really necessary?  If so could we bundle tgeneric_event_attr and uprobe_event_attr inside a single double-namespace block?  Otherwise, could we simply declare generic_event_attr and uprobe_event_attr like all the other functions in LinuxPerfUtils?

Reply:
Well, it's not necessary, it's to make these function "private" (static) in this .cpp file, but we could also make them like the other functions in LinuxPerfUtils. For now I moved them to a single namespace block.